File: android/app/src/main/java/ai/oriveo/community/core/navigation/AppRoute.kt
package ai.oriveo.community.core.navigation

import kotlinx.serialization.Serializable

@Serializable
enum class ProviderSetupEntryPoint {
    Onboarding,
    Providers,
    SkillEdit,
}

@Serializable
enum class ManualModelEntryContext {
    Onboarding,
    Providers,
    ProviderDetail,
    SkillEdit,
}

/** Navigation routes. */
sealed interface AppRoute {

    @Serializable data object Home : AppRoute
    @Serializable data object Providers : AppRoute
    @Serializable data object Settings : AppRoute

    @Serializable
    data class Chat(
        val conversationId: String? = null,
        val searchQuery: String? = null,
        val autoSend: Boolean = false,
        val focusMessageId: String? = null,
        val fromNoteId: String? = null,
    ) : AppRoute

    @Serializable data class FolderDetail(val folderID: String) : AppRoute
    @Serializable data object Notes : AppRoute
    @Serializable data class NoteDetail(val noteID: String) : AppRoute
    @Serializable data object Memory : AppRoute
    @Serializable data object Onboarding : AppRoute

    @Serializable
    data class ProviderSetup(
        val entryPoint: ProviderSetupEntryPoint = ProviderSetupEntryPoint.Onboarding,
        val preselectedKind: ai.oriveo.community.core.model.ProviderKind? = null,
    ) : AppRoute

    @Serializable
    data class ProviderDetail(
        val providerID: String,
        val entryPoint: ProviderSetupEntryPoint? = null,
    ) : AppRoute

    @Serializable
    data class ManualModelEntry(
        val providerID: String,
        val context: ManualModelEntryContext = ManualModelEntryContext.Providers,
    ) : AppRoute

    @Serializable
    data class RelaySetup(
        val entryPoint: ProviderSetupEntryPoint = ProviderSetupEntryPoint.Onboarding,
    ) : AppRoute

    @Serializable
    data class LocalComputeSetup(
        val entryPoint: ProviderSetupEntryPoint = ProviderSetupEntryPoint.Onboarding,
    ) : AppRoute

    @Serializable data object Backup : AppRoute
    @Serializable data object Skills : AppRoute
    @Serializable data class SkillEdit(val skillId: String? = null) : AppRoute
}

File: ios/Oriveo/Oriveo/Core/Routing/AppRoute.swift
import Foundation

enum ProviderSetupEntryPoint: String, Hashable {
    case welcome
    case providers
    case modelPicker
    case skillEdit

    var telemetryName: String {
        switch self {
        case .welcome: "onboarding"
        case .providers: "providers"
        case .modelPicker: "model_picker"
        case .skillEdit: "skill_edit"
        }
    }
}

enum ManualModelEntryContext: String, Hashable {
    case onboarding
    case providers
    case providerDetail
    case modelPicker
    case skillEdit
}

enum AppRoute: Hashable {
    case providerSetup(entryPoint: ProviderSetupEntryPoint, preselectedKind: ProviderKind? = nil)
    case manualModelEntry(providerID: UUID, context: ManualModelEntryContext)
    case providerDetail(providerID: UUID)
    case chat(conversationID: UUID?)
    case relaySetup(entryPoint: ProviderSetupEntryPoint)
    case localComputeSetup(entryPoint: ProviderSetupEntryPoint)
    case backup
    case folderDetail(folderID: UUID)
    case memory
    case skillsList
    case skillEdit(UUID?)
    case notesList
    case noteDetail(noteID: UUID)
}

/// One navigation policy shared by official providers, Relay, and Local compute setup.
enum ProviderSetupCompletionPolicy {
    /// Removes the complete setup subtree while preserving the route that launched it.
    /// A legacy deep link may enter Relay/Local without the parent providerSetup route, so both
    /// shapes are handled explicitly.
    static func returnedCallerPath(from path: [AppRoute]) -> [AppRoute] {
        if let setupRoot = path.lastIndex(where: { route in
            if case .providerSetup = route { return true }
            return false
        }) {
            return Array(path[..<setupRoot])
        }
        if let directSetup = path.lastIndex(where: { route in
            switch route {
            case .relaySetup, .localComputeSetup: return true
            default: return false
            }
        }) {
            return Array(path[..<directSetup])
        }
        return path
    }
}
